hawc/apps/lit/api.py

```diff
@@ -166,6 +169,64 @@ def tag_heatmap(self, request, pk):
         export = FlatExport(df=df, filename=f"df-{instance.id}")
         return Response(export)
 
+    @transaction.atomic()
+    @action(
+        detail=True, methods=("patch",), url_path="replace-hero",
+    )
+    def replace_hero(self, request, pk):
+        assessment = self.get_object()
+        body = json.loads(request.body)
+        replace = body.get("replace", list())
+        replace_unzipped = list(zip(*replace))
+        ref_ids = replace_unzipped[0]
+        hero_ids = replace_unzipped[1]
+
+        # make sure all references are HERO and in assessment
+        matching_references = assessment.references.filter(
+            id__in=ref_ids, identifiers__database=constants.HERO
+        )
+        if matching_references.count() != len(ref_ids):
+            raise exceptions.ValidationError("All references must be from selected assessment.")
+
+        # set hero ref
+        for index, (ref, hero) in enumerate(replace):
+            reference = models.Reference.objects.get(id=ref)
+            hero_identifier = reference.identifiers.get(database=constants.HERO)
+            setattr(hero_identifier, "unique_id", str(hero))
+            hero_identifier.save()
+
+        # update content
+        tasks.update_hero_content.apply(args=[hero_ids])
+
+        # update fields with content
+        for ref in ref_ids:
+            reference = models.Reference.objects.get(id=ref)
+            reference.update_from_hero_content()
+
+        return Response(status=status.HTTP_204_NO_CONTENT)
+
+    @transaction.atomic()
+    @action(
+        detail=True, methods=("patch",), url_path="update-reference-metadata-from-hero",
+    )
+    def update_reference_metadata_from_hero(self, request, pk):
+
+        # get all hero references from assessment
+        assessment = self.get_object()
+        references = assessment.references.filter(identifiers__database=constants.HERO)
+        reference_ids = set(references.values_list("id", flat=True))
+        identifiers = models.Identifiers.objects.filter(
+            references__in=reference_ids, database=constants.HERO
+        )
+        hero_ids = identifiers.values_list("unique_id", flat=True)
+        # update content of hero identifiers
+        tasks.update_hero_content.apply(args=[hero_ids])
+        # update fields from content
+        for reference in references:
```

Comment:
same as above, this needs to be chained in tasks, I think you could probably re-use the same new task that you'll be creating above.